src/Formats/ProtobufSerializer.cpp

```diff
@@ -1969,6 +1971,89 @@ namespace
     };
 
 
+    /// Wraps a structute (field, Message, etc) which is a member of OneOf (protobuf union)
+    class ProtobufSerializerOneOf : public ProtobufSerializer
+    {
+    public:
+        explicit ProtobufSerializerOneOf(std::unique_ptr<ProtobufSerializer> nested_serializer_, std::string_view one_of_column_name_, size_t column_idx_, int field_tag_)
+            : nested_serializer(std::move(nested_serializer_))
+            , one_of_column_name(one_of_column_name_)
+            , column_idx(column_idx_)
+            , field_tag(field_tag_)
+        {
+        }
+
+        void setColumns(const ColumnPtr * columns, size_t num_columns) override
+        {
+            assert(num_columns > column_idx);
+
+            Columns cols;
+            cols.reserve(num_columns - 1);
+            for (size_t i : collections::range(num_columns))
+            {
+                if (i == column_idx)
+                {
+                    presence_column = columns[column_idx]->assumeMutable();
+                }
+                else
+                {
+                    cols.push_back(columns[i]->getPtr());
+                }
+            }
+            nested_serializer->setColumns(cols.data(), cols.size());
+        }
+
+        void setColumns(const MutableColumnPtr * columns, size_t num_columns) override
+        {
+            Columns cols;
+            cols.reserve(num_columns);
+            for (size_t i : collections::range(num_columns))
+                cols.push_back(columns[i]->getPtr());
+            setColumns(cols.data(), cols.size());
+        }
+
+        void writeRow([[maybe_unused]] size_t row_num) override
+        {
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "OneOf is not implemented for writes");
+            // Should be as simple as
+            //   if (static_cast<int>(presence_column->getInt(row_num)) == field_tag)
+            //       nested_serializer->writeRow(row_num);
+            // but disabled so far
+        }
+
+        void readRow(size_t row_num) override
+        {
+            if (presence_column)
+            {
+                presence_column->insert(field_tag);
+            }
+            nested_serializer->readRow(row_num);
+        }
+
+        void insertDefaults(size_t row_num) override
+        {
+            if (row_num >= presence_column->size())
+                presence_column->insert(0);
+
+            nested_serializer->insertDefaults(row_num);
+        }
+
+        void describeTree(WriteBuffer & out, size_t indent) const override
+        {
+            writeIndent(out, indent) << "ProtobufSerializerOneOf " << one_of_column_name << ", idx " << column_idx << "->\n";
+            nested_serializer->describeTree(out, indent + 1);
+        }
+
+    private:
+        const std::unique_ptr<ProtobufSerializer> nested_serializer;
+        ColumnPtr column; // ???
+        std::string_view one_of_column_name;  // Is it safe?
+        size_t column_idx;
```

Comment:
Is it an index of presence column? If yes, please, rename the field to `presence_column_idx`

Reply:
Yes, will rename to presence_column_idx.